File: mobile/lib/features/auth/data/auth_api.dart
import 'package:dio/dio.dart';

class AuthApiException implements Exception {
  AuthApiException(this.message, {this.fieldErrors = const {}});

  final String message;
  final Map<String, List<String>> fieldErrors;
}

class AuthApi {
  AuthApi(this._dio);

  final Dio _dio;

  Future<Map<String, dynamic>> register({
    required String name,
    required String email,
    required String password,
    required String passwordConfirmation,
  }) async {
    final response = await _post('/register', {
      'name': name,
      'email': email,
      'password': password,
      'password_confirmation': passwordConfirmation,
    });
    return response['data'] as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> login({required String email, required String password}) async {
    final response = await _post('/login', {'email': email, 'password': password});
    return response['data'] as Map<String, dynamic>;
  }

  Future<void> logout() => _post('/logout', const {});

  Future<void> forgotPassword(String email) => _post('/forgot-password', {'email': email});

  Future<void> resetPassword({
    required String token,
    required String email,
    required String password,
    required String passwordConfirmation,
  }) =>
      _post('/reset-password', {
        'token': token,
        'email': email,
        'password': password,
        'password_confirmation': passwordConfirmation,
      });

  Future<Map<String, dynamic>> _post(String path, Map<String, dynamic> data) async {
    try {
      final response = await _dio.post<Map<String, dynamic>>(path, data: data);
      return response.data ?? const {};
    } on DioException catch (e) {
      final body = e.response?.data;
      if (body is Map<String, dynamic>) {
        final errors = body['errors'];
        throw AuthApiException(
          body['message'] as String? ?? 'Une erreur est survenue.',
          fieldErrors: errors is Map
              ? errors.map(
                  (key, value) => MapEntry(
                    key as String,
                    (value as List).map((e) => e.toString()).toList(),
                  ),
                )
              : const {},
        );
      }
      throw AuthApiException('Impossible de contacter le serveur. Vérifiez votre connexion Internet.');
    }
  }
}

File: appReact/src/features/auth/data/authApi.ts
import { AxiosInstance, AxiosError } from 'axios';

export class AuthApiError extends Error {
  fieldErrors: Record<string, string[]>;

  constructor(message: string, fieldErrors: Record<string, string[]> = {}) {
    super(message);
    this.name = 'AuthApiError';
    this.fieldErrors = fieldErrors;
  }
}

interface AuthSessionPayload {
  token: string;
  user: any;
}

export class AuthApi {
  constructor(private http: AxiosInstance) {}

  register(params: { name: string; email: string; password: string; passwordConfirmation: string }): Promise<AuthSessionPayload> {
    return this.post('/register', {
      name: params.name,
      email: params.email,
      password: params.password,
      password_confirmation: params.passwordConfirmation,
    });
  }

  login(params: { email: string; password: string }): Promise<AuthSessionPayload> {
    return this.post('/login', { email: params.email, password: params.password });
  }

  async logout(): Promise<void> {
    await this.post('/logout', {});
  }

  async forgotPassword(email: string): Promise<void> {
    await this.post('/forgot-password', { email });
  }

  async resetPassword(params: { token: string; email: string; password: string; passwordConfirmation: string }): Promise<void> {
    await this.post('/reset-password', {
      token: params.token,
      email: params.email,
      password: params.password,
      password_confirmation: params.passwordConfirmation,
    });
  }

  private async post(path: string, data: Record<string, unknown>): Promise<any> {
    try {
      const response = await this.http.post(path, data);
      return response.data?.data;
    } catch (err) {
      const error = err as AxiosError<{ message?: string; errors?: Record<string, string[]> }>;
      const body = error.response?.data;
      if (body) {
        throw new AuthApiError(body.message ?? 'Une erreur est survenue.', body.errors ?? {});
      }
      throw new AuthApiError('Impossible de contacter le serveur. Vérifiez votre connexion Internet.');
    }
  }
}
